### PR TITLE
(Wallet) Implement recovery screens

### PR DESCRIPTION
**Pull request by simoarpe**
## Resolves https://github.com/brave/brave-browser/issues/40168

This PR implements new layouts and new logic for recovery screens that cover many parts of the onboarding flow.

### Recovery Phrase

| Light Theme   |     Dark Theme      | 
|:----------:|:-------------:|
| Screenshot_1722511123| Screenshot_1722511185 | 

The recovery phrase is shown when a new Wallet is created and the UI is presented with a blurred component that will be unblurred only upon user confirmation.

### Recovery Phrase Verification in 3 Steps

| Light Theme   |     Dark Theme      | 
|:----------:|:-------------:|
| Screenshot_1722511585| Screenshot_1722511362 | 

The recovery phrase verification is now in three steps and will ask to confirm three words picked at a random position.
For security reasons these steps won't allow back navigation and skipping them will consider the backup process as not copmpleted, so the user will have the option to complete the verification later on. 

### Pre-existing Wallet Restoration

| Light Theme   |     Dark Theme      | 
|:----------:|:-------------:|
| Screenshot_1722511669| Screenshot_1722511699 | 

The new Wallet restoration screen is the most complex screen of this PR. It accepts a mnemonic phrase made by 12 or 24 words. It can be copied over from clipboard, spit, formatted and pasted nicely in each edit text. The edit text items have been customized to accept the pasted text (See `PasteEditText` component). The layout can switch from 12 and 24 words, and also has a button to mask and unmask the words. The continue button is responsive and is enabled ONLY when all the items have been filled.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Recovery Phrase
- Clean previous Wallet data, if any
- Navigate to Wallet section to trigger onboarding flow
- Select new Wallet and follow onboarding process untill recovery phrase screen is present
- Verify recovery phrase is initially blurred
- Tap on the button "View my recovery phrase"
- Observe the recovery phrase is now visible
- Tap on Android home button to put app in background
- Access the app again
- Observe the recovery phrase is blurred again
- Tap on the button "View my recovery phrase"
- Observe the button has changed its name
- Observe a new text view is displayed
- Tap on the new text view
- Verify the recovery phrase has been copied correctly
- Tap on the skip button
- Observe a new dialog showing up
 
### Recovery Phrase Verification in 3 Steps
- Clean previous Wallet data, if any
- Navigate to Wallet section to trigger onboarding flow
- Select new Wallet and follow onboarding process untill the first recovery phrase verification screen is present
- Observe the continue button is disabled
- Fill the edit text with the word "test" and tap the continue button
- Observe an error message is shown
- Fill the edit text with the right corresponding word
- Observe the screen will ask for a second verification
- Fill the edit text with the right corresponding word
- Observe the screen will ask for a third verification

### Pre-existing Wallet Restoration
- Clean previous Wallet data, if any
- Navigate to Wallet section to trigger onboarding flow
- Select an existing Wallet to restore and follow onboarding process untill the recovery phrase input screen is presented
- Tap on the 24-word edit text
- Observe the layout changes to 24 input words
- Tap again on the 12-word edit text
- Observe the layout changes back to 12 input words
- Paste a valid recovery phrase in a random item
- Observe the recovere phrase is nicely pasted in each edit text
- Tap on the eye icon
- Observe the words are visible
- Observe the continue button is responsive and enabled ONLY when all the items are filled

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java:179]: This code was moved to a base class to avoid duplication and left untouched, I thought it was already vetted by security team. 
I've filed a new GitHub issue to investigate if it's a false positive from the code scanning tool: https://github.com/brave/brave-browser/issues/40188

Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java:178]: Fingeprint authentication algorithm has a security review:
* https://github.com/brave/brave-core/pull/11298 
* https://github.com/brave/reviews/issues/677

File: browser/ui/android/strings/android_brave_strings.grd
Comment (SergeyZhukovsky) [inline browser/ui/android/strings/android_brave_strings.grd:2009]: `’` -> `'`

File: android/java/org/chromium/chrome/browser/crypto_wallet/model/OnboardingViewModel.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/model/OnboardingViewModel.java:54]: do we catch that exception on places where we call `requireRecoveryPhrase`?

File: browser/ui/android/strings/android_brave_strings.grd
Comment (mkarolin) [inline browser/ui/android/strings/android_brave_strings.grd:976]: I believe these should be `12-word` and `24-word`. cc: @rmcfadden3

Comment (rmcfadden3) [reply browser/ui/android/strings/android_brave_strings.grd:976]: @mkarolin — yes! should be singular. `12-word` and `24-word`

Comment (simoarpe) [reply browser/ui/android/strings/android_brave_strings.grd]: 👍 Fixed by commit [`f72134b` (#24943)](https://github.com/brave/brave-core/pull/24943/commits/f72134ba2bb49d5e55b9c43e7dae761eb6f6a13d)

File: android/java/org/chromium/chrome/browser/crypto_wallet/model/OnboardingViewModel.java
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/model/OnboardingViewModel.java:54]: No we don't. It was an extra check I've introduced just to catch early an error that we may introduce during development/debugging. So better switch it to an assertion.

Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/model/OnboardingViewModel.java]: Fixed in commit [`a01af24` (#24943)](https://github.com/brave/brave-core/pull/24943/commits/a01af24cd545c8bf0c04b9e106c45d8f124f9a2c)

File: android/java/org/chromium/chrome/browser/crypto_wallet/model/OnboardingViewModel.java
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/model/OnboardingViewModel.java]: Fixed in commit [`a01af24` (#24943)](https://github.com/brave/brave-core/pull/24943/commits/a01af24cd545c8bf0c04b9e106c45d8f124f9a2c)

**Review by mkarolin — APPROVED** [review]
`strings`++

File: android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java
Comment (deeppandya) [inline android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java:425]: should we use LiveDataUtil.observeOnce here ?

Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/app/domain/KeyringModel.java:425]: Unfortunately it didn't work here as I needed to observe the `createAccountsLiveData` not only once but for the entire time the accounts are emitted by `mAccountInfos` (a few lines below). When the number of emitted accounts matches the size of `createAccounts` then and only then we can suspend the observer.

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Comment by diracdeltas** [comment]
adding @stoletheminerals  for security review

File: android/java/res/layout/fragment_recovery_phrase.xml
Comment (deeppandya) [inline android/java/res/layout/fragment_recovery_phrase.xml:29]: can we avoid using textStyle here ? it should be handled with font weight.

**Review by deeppandya — APPROVED** [review]
LGTM with minor comment if we can avoid using textStyle

File: android/java/res/layout/fragment_recovery_phrase.xml
Comment (simoarpe) [reply android/java/res/layout/fragment_recovery_phrase.xml:29]: @deeppandya , I went through all the places where `bold` is used. Would it be possible to create the bold equivalent for these styles? 

`Heading1` -> `Heading1Bold`
`Heading3` -> `Heading3Bold`
`DefaultRegular` -> `DefaultBold`
`XSmallSemibold` -> `XSmallBold`

Each of the new styles above can inherit all the attributes from their non-bold parent and just add `android:textStyle="bold"`. 
I can help with this if you want, then we merge, rebase this branch on top of master and remove all the bold attributes from all the components. How does that sound?

**Comment by stoletheminerals** [comment]
> 

Can you make the blurred words to be the same length? See: https://github.com/brave/internal/issues/915

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java
Comment (stoletheminerals) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java:178]: Is this a part of the onboarding flow? If yes than it's probably a false positive. At this stage we only set up the biometric authentication. Once it was set up, we will be using biometrics to access private key in the keychain, but we don't need to do it at the onboarding stage.

Comment (stoletheminerals) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java:178]: oh nevermind, seems like we have an actual issue here. `authenticationResult` should be returning a secret that will decrypt wallet's password

**Comment by simoarpe** [comment]
@stoletheminerals, the words have now the same length when blurred. Feedback addressed by commit [`bcf4f04` (#24943)](https://github.com/brave/brave-core/pull/24943/commits/bcf4f04b3a660216c3429ad14588e9e463ad5d32).

Preview:
Screenshot_1722870796

File: android/java/res/layout/fragment_recovery_phrase.xml
Comment (deeppandya) [reply android/java/res/layout/fragment_recovery_phrase.xml:29]: @aguscruiz WDYT ? bold style could be handled with font weights.

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java:178]: Created a PR to fix this: https://github.com/brave/brave-core/pull/25104

**Comment by thypon** [comment]
@simoarpe are the content of the blurred n-words something like `******` or is the real content + padding?

It is possible to deblur with ML methods, since even if blurred, each letter follows a similar pattern which can be caught and translated back into a real letter?

**Comment by simoarpe** [comment]
@thypon, the blurred content is made just from a static placeholder instead of actual words that is for `XXXXXXX` for all the items. I've used Xs insteads or *s as it gives a nicer effect when blurred.

https://github.com/brave/brave-core/blob/87ee38987570273b8a54a3c8bdd210ff33f8a3b3/android/java/org/chromium/chrome/browser/crypto_wallet/adapters/RecoveryPhraseAdapter.java#L28 

blur

File: android/java/org/chromium/chrome/browser/crypto_wallet/adapters/RecoveryPhraseAdapter.java
Comment (thypon) [inline android/java/org/chromium/chrome/browser/crypto_wallet/adapters/RecoveryPhraseAdapter.java:74]: Add a comment here

```
// XXX SECURITY: NEVER BLUR REAL RECOVERY PHRASE
//                             !!! USE A PLACEHOLDER !!!
```

**Review by thypon — CHANGES_REQUESTED** [review]

File: android/java/org/chromium/chrome/browser/crypto_wallet/adapters/RecoveryPhraseAdapter.java
Comment (simoarpe) [reply android/java/org/chromium/chrome/browser/crypto_wallet/adapters/RecoveryPhraseAdapter.java:76]: 👍

**Review by thypon — APPROVED** [review]

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java
Comment (bridiver) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/BaseWalletNextPageFragment.java:231]: I don't think we're supposed to use this on android cc @SergeyZhukovsky

**Pull request merged**
